src/modules/0.0.21/aws_codebuild/entity/project.ts

```diff
@@ -0,0 +1,111 @@
+import { Column, Entity, PrimaryColumn, ManyToOne, JoinColumn } from 'typeorm';
+
+import { cloudId } from '../../../../services/cloud-id';
+import { Role } from '../../aws_iam/entity/role';
+
+export enum SourceType {
+  GITHUB = 'GITHUB',
+  // CODECOMMIT = 'CODECOMMIT',
+  // CODEPIPELINE = 'CODEPIPELINE',
+  // S3 = 'S3',
+  // BITBUCKET = 'BITBUCKET',
+  // GITHUB_ENTERPRISE = 'GITHUB_ENTERPRISE',
+  // NO_SOURCE = 'NO_SOURCE'
+}
+
+export enum EnvironmentVariableType {
+  PLAINTEXT = 'PLAINTEXT',
+  // PARAMETER_STORE = 'PARAMETER_STORE',
+  // SECRETS_MANAGER = 'SECRETS_MANAGER',
+}
+
+// https://docs.aws.amazon.com/codebuild/latest/userguide/build-env-ref-compute-types.html
+export enum ComputeType {
+  BUILD_GENERAL1_SMALL = 'BUILD_GENERAL1_SMALL',
+  BUILD_GENERAL1_MEDIUM = 'BUILD_GENERAL1_MEDIUM',
+  BUILD_GENERAL1_LARGE = 'BUILD_GENERAL1_LARGE',
+  BUILD_GENERAL1_2XLARGE = 'BUILD_GENERAL1_2XLARGE',
+}
+export enum EnvironmentType {
+  ARM_CONTAINER = 'ARM_CONTAINER',
+  LINUX_CONTAINER = 'LINUX_CONTAINER',
+  LINUX_GPU_CONTAINER = 'LINUX_GPU_CONTAINER',
+  WINDOWS_CONTAINER = 'WINDOWS_CONTAINER',
+  WINDOWS_SERVER_2019_CONTAINER = 'WINDOWS_SERVER_2019_CONTAINER',
+}
+
+// TODO support buildspec file in repo
+@Entity()
+export class CodebuildProject {
+  @PrimaryColumn()
+  @cloudId
+  projectName: string;
+
+  @Column({
+    nullable: true,
+  })
+  arn?: string;
+
+  // https://docs.aws.amazon.com/codebuild/latest/userguide/build-spec-ref.html
+  @Column({
+    type: 'text',
+    nullable: true,
+  })
+  buildSpec?: string;
+
+  // For a GitHub repository, the HTTPS clone URL with the source code location
+  @Column()
+  sourceLocation: string;
+
+  @Column({
+    default: 'aws/codebuild/standard:6.0',
+  })
+  image: string;
+
+  @ManyToOne(() => Role, {
+    eager: true,
+  })
+  @JoinColumn({
+    name: 'service_role_name',
+  })
+  serviceRole: Role;
+
+  @Column({
+    type: 'enum',
+    enum: ComputeType,
+    default: ComputeType.BUILD_GENERAL1_SMALL,
+  })
+  computeType: ComputeType;
+
+  @Column({
+    type: 'enum',
+    enum: EnvironmentType,
+    default: EnvironmentType.LINUX_CONTAINER,
+  })
+  environmentType: EnvironmentType;
+
+  // https://docs.aws.amazon.com/codebuild/latest/userguide/sample-source-version.html
+  @Column({
+    nullable: true,
+  })
+  sourceVersion?: string;
+
+  @Column({
+    type: 'enum',
+    enum: SourceType,
+  })
+  sourceType: SourceType;
+
+  @Column({
+    type: 'simple-json',
+    nullable: true,
+  })
+  environmentVariables?: [{ name: string; value: string; type: EnvironmentVariableType }];
```

Comment:
We don't actually have real type-safety here. The user could insert whatever arbitrary JSON they want here but Typescript will tell us it matches the exact type you have here.

This is a general problem we need to figure out, but not excited to see more tech debt getting piled up here.

Also, since there's only one type of environment variable supported here, why not just make it `{ [key: string]: string, }` as the type, which is much more natural to work with?

Reply:
well, eventually I was hoping it would support secrets manager so I would rather leave the format as-is